// yggtorrent-api/index.js
var request = require('request');
var cheerio = require('cheerio');
const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const axios = require('axios');
const tunnel = require('tunnel');
const settings = require('../../../settings');
var jar = request.jar();
request.defaults({followAllRedirects: true});
const https = require('https');
class Ygg {
    constructor(config) {
        this.host = config.host;
        this.searchhost = config.searchhost;
        this.username = config.username;
        this.password = config.password;
        this.proxyOptions = {
            proxy: {
                host: '51.254.121.123',
                port: 8088,
                // Add any other proxy options you need
            }
        };
    }

    login(callback) {
        const self = this; // Stocker une référence à "this" pour l'utiliser dans la fonction de rappel

        request({
            method: 'POST',
            url: this.host+'/user/login',
            headers: {
                'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.120 Safari/537.36',
                'x-requested-with': 'XMLHttpRequest',
            },
            formData: {
                'id': this.username,
                'pass': this.password,
            },
            jar: jar
        }, (err, response, body) => {
            if (err) return callback(err);
            if (response.statusCode / 100 >= 4) {
                var error = new Error('Bad status code while login : '+response.statusCode+'. Bad username/password ?')
                error.body = body;
                return callback(error);
            }
            self.jar = jar;
            callback(err, body);
        });
    }

    getRatio(callback) {
        request({
            method: 'GET',
            url: this.host + '/user/ajax_usermenu',
            headers: {
                'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.120 Safari/537.36',
                'x-requested-with': 'XMLHttpRequest',
            },
            jar: jar,
            json: true,
        }, (err, response, body) => {
            if (err) return callback(err);
            if (response.statusCode / 100 >= 4) {
                var error = new Error('Bad status code getting ratio : ' + response.statusCode);
                error.body = body;
                return callback(error);
            }

            body = body.html;
            
            // console.log(body);
            var $ = cheerio.load(body);
            var ratio = body.match(/Ratio : ([0-9\.]+)/)
            var results = {
                upload: $('.ico_upload').parent().text(),
                download: $('.ico_download').parent().text(),
                ratio: ratio[1],
            };

            callback(err, results);
        });
    }

    search(name, callback) {
        request({
            method: 'GET',
            url: this.searchhost + '/engine/search',
            qs: {
                name: name,
                do: 'search',
                category: '2145',
            },
            headers: {
                'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/77.0.3865.120 Safari/537.36',
            },
            
            tunnel: tunnel.httpsOverHttp(this.proxyOptions)
        }, (err, response, body) => {
            if (err) return callback(err);
            if (response.statusCode / 100 >= 4) {
                var error = new Error('Bad status code while searching : ' + response.statusCode)
                error.body = body;
                return callback(error);
            }

            var $ = cheerio.load(body);

            var results = [];
            $('.table-responsive.results tbody tr').each((i, tr) => {
                results.push({
                    // line: $(tr).text(),
                    url: $(tr).find('#torrent_name').attr('href'),
                    name: $(tr).find('#torrent_name').text().trim(),
                    size: $($(tr).find('td')[5]).text(),
                    downloadurl: this.searchhost + '/engine/download_torrent?id=' + $(tr).find('#get_nfo').attr('target'),
                });
            })
            callback(err, results);
        });
    }
    downloadLink(link, callback) {
        // Générer un nom de fichier aléatoire en utilisant un identifiant unique
        const randomFileName = crypto.randomBytes(10).toString('hex') + '.torrent';
        
        // Chemin de destination complet en utilisant le nom de fichier aléatoire
        const outputFilePath = path.join('D:/download', randomFileName);
        const yggCookie = this.jar.getCookieString(settings.CONSTANT_URLYGG);
        axios({
            method: 'get',
            url: link,
            responseType: 'stream',
            httpsAgent: new https.Agent({ rejectUnauthorized: false }),
            headers: {
                Cookie: yggCookie
            }
        })
        .then(response => {
            const outputStream = fs.createWriteStream(outputFilePath);
            response.data.pipe(outputStream);
    
            outputStream.on('finish', () => {
                callback(null, 'Téléchargement réussi');
            });
        })
        .catch(error => {
            console.error('Une erreur s\'est produite lors du téléchargement :', error);
            callback(error, null); // Appel du callback avec l'erreur en cas d'échec
        });
    }
}
module.exports = Ygg;
